File: backend/src/routes/crops.ts
import { Router, Request, Response } from 'express';
import axios from 'axios';
import { prisma } from '../prisma';

const router = Router();

/**
 * POST /api/crops/recommend
 * Body: { soil, ph, rainfall, temp, region, season, nitrogen, water }
 *
 * Forwards to Python AI service and returns:
 * { crops: [{ emoji, name, score, why }], summary }
 */
router.post('/recommend', async (req: Request, res: Response) => {
  try {
    const { soil, ph, rainfall, temp, region, season, nitrogen, water } = req.body;

    const aiServiceUrl = process.env.AI_SERVICE_URL ?? 'http://localhost:8000';
    const aiResponse = await axios.post<AiCropResponse>(
      `${aiServiceUrl}/crops/recommend`,
      { soil, ph, rainfall, temp, region, season, nitrogen, water },
      { timeout: 30_000 },
    );

    res.json(aiResponse.data);
  } catch (err) {
    console.error('[crops/recommend]', err);
    res.status(502).json({ error: 'Crop recommendation service unavailable' });
  }
});

/**
 * GET /api/crops/fields
 * Returns fields with current crop info (for dropdown population in UI).
 */
router.get('/fields', async (_req: Request, res: Response) => {
  try {
    const fields = await prisma.field.findMany({
      select: {
        id: true,
        name: true,
        cropType: true,
        cropSeason: true,
        soilType: true,
        soilPh: true,
        areaHa: true,
      },
      orderBy: { name: 'asc' },
    });
    res.json({ fields });
  } catch (err) {
    console.error('[crops/fields]', err);
    res.status(500).json({ error: 'Failed to fetch fields' });
  }
});

interface AiCropResponse {
  crops: Array<{
    emoji: string;
    name: string;
    score: number;
    why: string;
  }>;
  summary: string;
}

export default router;

File: ai-service/app/routers/crops.py
"""
Crop recommendation router.
POST /crops/recommend — accepts soil/climate parameters
"""

import logging
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)
router = APIRouter()


class CropRecommendRequest(BaseModel):
    soil:     str   = Field("loamy",  description="Soil type: loamy, clay, sandy, silt, black, red")
    ph:       float = Field(6.5,      ge=3.0, le=10.0, description="Soil pH")
    rainfall: float = Field(800.0,    ge=0,   description="Annual rainfall in mm")
    temp:     float = Field(25.0,     description="Average temperature in °C")
    region:   str   = Field("Punjab", description="State / region in India")
    season:   str   = Field("rabi",   description="Season: rabi, kharif, zaid")
    nitrogen: float = Field(50.0,     ge=0,   description="Soil nitrogen (kg/ha)")
    water:    str   = Field("medium", description="Water availability: low, medium, high")


class CropItem(BaseModel):
    emoji: str
    name:  str
    score: float
    why:   str


class CropRecommendResponse(BaseModel):
    crops:   list[CropItem]
    summary: str


@router.post("/recommend", response_model=CropRecommendResponse)
async def recommend_crops(body: CropRecommendRequest, request: Request):
    """
    Returns up to 6 ranked crop recommendations.

    The response JSON exactly matches the shape expected by the
    Node.js backend (and ultimately the frontend MOCK.cropRec format):
    {
      "crops": [
        { "emoji": "🌾", "name": "Wheat", "score": 95, "why": "..." },
        ...
      ],
      "summary": "..."
    }
    """
    model = request.app.state.crop_model
    try:
        crops = model.recommend(body.model_dump())
    except Exception as exc:
        logger.exception("Crop model inference failed: %s", exc)
        raise HTTPException(status_code=500, detail="Model inference error")

    if not crops:
        raise HTTPException(status_code=500, detail="No recommendations generated")

    top = crops[0]["name"]
    summary = (
        f"Based on {body.soil} soil (pH {body.ph}), {body.rainfall} mm rainfall, "
        f"and {body.season} season in {body.region}, "
        f"{top} is your best option with {crops[0]['score']}% match score."
    )

    return CropRecommendResponse(crops=crops, summary=summary)
